src/dots.c

```diff
@@ -0,0 +1,502 @@
+    sexp* expr = dot_get_expr(elt);
+    sexp* env = dot_get_env(elt);
+
+    // Unquoting rearranges expressions
+    expr = KEEP(r_duplicate(expr, false));
+
+    if (capture_info->unquote_names && r_is_call(expr, ":=")) {
+      sexp* name = def_unquote_name(expr, env);
+
+      if (dots_names == r_null) {
+        dots_names = KEEP(r_new_vector(r_type_character, n));
+        r_push_names(dots, dots_names);
+        FREE(1);
+      }
+
+      if (r_chr_has_empty_string_at(dots_names, i)) {
+        r_chr_poke(dots_names, i, name);
+      } else {
+        r_abort("Can't supply both `=` and `:=`");
+      }
+      expr = r_node_cadr(r_node_cdr(expr));
+    }
+
+    struct expansion_info info = which_expansion_op(expr);
+    enum dots_expansion_op dots_op = info.op + (N_EXPANSION_OPS * capture_info->type);
+
+    // Ignore empty arguments
+    if (expr == r_missing_sym
+        && (dots_names == r_null || r_chr_has_empty_string_at(dots_names, i))
+        && should_ignore(capture_info->ignore_empty, i, n)) {
+      capture_info->needs_expansion = true;
+      r_list_poke(dots, i, empty_spliced_list());
+      FREE(1);
+      continue;
+    }
+
+    switch (dots_op) {
+    case OP_EXPR_NONE:
+    case OP_EXPR_UQ:
+    case OP_EXPR_UQE:
+      expr = call_interp_impl(expr, env, info);
+      capture_info->count += 1;
+      break;
+    case OP_EXPR_UQS:
+      capture_info->needs_expansion = true;
+      expr = dots_big_bang(capture_info, info.operand, env, false);
+      break;
+    case OP_QUO_NONE:
+    case OP_QUO_UQ:
+    case OP_QUO_UQE: {
+      expr = KEEP(call_interp_impl(expr, env, info));
+      expr = forward_quosure(expr, env);
+      FREE(1);
+      capture_info->count += 1;
+      break;
+    }
+    case OP_QUO_UQS: {
+      capture_info->needs_expansion = true;
+      expr = dots_big_bang(capture_info, info.operand, env, true);
+      break;
+    }
+    case OP_VALUE_NONE:
+      if (expr == r_missing_sym) {
+        r_abort("Argument %d is empty", i + 1);
+      }
+      expr = r_eval(expr, env);
+      if (r_inherits(expr, "spliced")) {
+        capture_info->needs_expansion = true;
+      }
+      capture_info->count += 1;
+      break;
+    case OP_VALUE_UQ:
+    case OP_VALUE_UQE:
+      r_abort("Can't use `!!` in a non-quoting function");
+    case OP_VALUE_UQS: {
+      expr = KEEP(r_eval(info.operand, env));
+      capture_info->needs_expansion = true;
+      if (expr == r_null) {
+        expr = empty_spliced_list();
+      } else {
+        expr = set_value_spliced(expr);
+        capture_info->count += 1;
+      }
+      FREE(1);
+      break;
+    }}
+
+    r_list_poke(dots, i, expr);
+    FREE(1);
+  }
+
+  return dots;
+}
+
+
+static int match_ignore_empty_arg(sexp* ignore_empty) {
+  if (!r_is_character(ignore_empty) || r_length(ignore_empty) == 0) {
+    r_abort("`.ignore_empty` must be a character vector");
+  }
+  const char* arg = r_c_string(ignore_empty);
+  switch(arg[0]) {
```

Comment:
This seems excessively efficient 😉 

(this is ok if `.ignore_empty = ""` because you can switch on null?)